### ISSUE TITLE
[Proposal] Cast query builder result sets to their db native type

### ISSUE THREAD

**Issue by phroggyy** [laravel/internals#244]
TL;DR: Should Laravel automatically cast column values to the type used in the database?

---

Due to how PDO works, the type that is returned from a PDO statement for a given column is always a string (and, well, that string can be somewhat unpredictable too). However, with [`PDOStatement::getColumnMeta`](http://us3.php.net/manual/en/pdostatement.getcolumnmeta.php), it's actually possible to find out the type of given columns. This would allow us to correctly cast to whatever the database is setup to use, which might make some things easier when dealing with strict checks. Naturally, we already have `$casts`, but it'd be pretty smooth if it _Just Worked™_.

It should also be noted that the string returns is actually a driver issue, and not a PHP one, so for mysql folks, using mysqlnd _should_ actually work as long as Laravel applies:

``` php
$pdo->setAttribute(PDO::ATTR_STRINGIFY_FETCHES, false);
$pdo->setAttribute(PDO::ATTR_EMULATE_PREPARES, false);
```

That said, it'd be nice if we could cast to native for everyone.

Thoughts?

**Comment by michaeldyrynda** [comment]
I wonder how this works under the hood. Does it hit the database for each column, or is all the data retrieved at once? I'd not use it if the former, as that's implicit query overhead you don't need in an application, particularly if you're not worried about types.

Also, you'll get inconsistent behaviour if you're using SQLite as it is untyped, which could run into testing issues perhaps (or more), which is why the `$casts` property is a reasonable middle ground IMO.

I'd be more inclined to look at the `setAttribute` option, rather than trying to make calls fetch data after the fact, if that is going to behave more consistently.
